### PR TITLE
[SPARK-6752][STREAMING][Revised] Allow StreamingContext to be recreated from checkpoint and existing SparkContext

### PR DESCRIPTION
**Pull request by tdas**
This is a revision of the earlier version (see #5773) that passed the active SparkContext explicitly through a new set of Java and Scala API. The drawbacks are.
- Hard to implement in python. 
- New API introduced. This is even more confusing since we are introducing getActiveOrCreate in SPARK-7553

Furthermore, there is now a direct way get an existing active SparkContext or create a new on - SparkContext.getOrCreate(conf). Its better to use this to get the SparkContext rather than have a new API to explicitly pass the context.

So in this PR I have
- Removed the new versions of StreamingContext.getOrCreate() which took SparkContext
- Added the ability to pick up existing SparkContext when the StreamingContext tries to create a SparkContext.

### CODE REVIEW

File: streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala [hunk 1]
```diff
@@ -418,76 +418,16 @@ class StreamingContextSuite extends FunSuite with BeforeAndAfter with Timeouts w
...
-      assert(newContextCreated, "new context not created")
-      assert(ssc.sparkContext === sc, "new StreamingContext does not use existing SparkContext")
-    }
-
-    val checkpointPath = createValidCheckpoint()
-
-    // StreamingContext.getOrCreate should recover context with checkpoint path
-    testGetOrCreate {
-      ssc = StreamingContext.getOrCreate(checkpointPath, creatingFunction _, sc)
+      sc = new SparkContext(conf)
```
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala]: Here the whole unit test to test the deleted API has been replaced by a sub-unit-test that tests older API's ability to use an existing SparkContext.

**Comment by tdas** [comment]
@zsxwing

**Comment by zsxwing** [comment]
LGTM

**Comment by tdas** [comment]
@JoshRosen Since you took a look at the original PR #5428, could you take a look at this?
Also @harishreedharan.

**Comment by tdas** [comment]
test this please.

**Comment by tdas** [comment]
test this please

**Comment by JoshRosen** [comment]
This seems fine to me.  My only comment would be whether we need to document this behavior anywhere or whether this behavior change might cause confusion for users.  I don't think it should cause issues, though, since the old code would have just thrown a confusing "multiple active SparkContexts in the same JVM" message in these scenarios.

**Comment by tdas** [comment]
Yeah, this would not have been a possible code path before due to that
error. So I dont think this changes anything for anyone.  Alright, I am
merging this. Thanks all!

**Comment by danielwegener** [comment]
Just curious. How are we supposed to create a `StreamingContext` from a given `SparkContext` and a directory if we don't know if the directory exists (which matches the dropped _confusing_ getOrCreate-overloading)?

**Comment by tdas** [comment]
@danielwegener In the creating function that you pass on to getOrCreate, you can do 
```
def creatingFunc(): StreamingContext = {
    val ssc = new StreamingContext(SparkContet.getOrCreate(), batchInterval)
     OR
    val ssc = new StreamingContext(referenceToExistingSparkContext, batchInterval)
}
```

**Pull request closed without merging**